lib/evaluate/characteristics.h

```diff
@@ -89,13 +102,34 @@ class TypeAndShape {
 
 private:
   void AcquireShape(const semantics::ObjectEntityDetails &);
+  void AcquireLEN();
 
 protected:
   DynamicType type_;
+  std::optional<Expr<SomeInteger>> LEN_;
```

Comment:
A `DynamicType` instance can't contain an arbitrary `CHARACTER` length expression, but `TypeAndShape` can.  This allows `TypeAndShape` to fully characterize `CHARACTER` expressions as well as symbols, so that actual arguments as in e.g. `CALL FOO(X//Y)` can be represented.